### PR TITLE
DM-21950: Update validate_drp to work with different external calibrations

### PR DESCRIPTION
**Pull request by erykoff**
Update validate_drp so that external photometric and astrometric corrections are separated, and so that fgcmcal can be used in place of jointcal for photometry.

### CODE REVIEW

File: python/lsst/validate/drp/matchedVisitMetricsTask.py [hunk 1]
```diff
@@ -30,8 +33,19 @@ def getTargetList(parsedCmd, **kwargs):
...
     def __call__(self, args):
+        exitStatus = 0
+
         task = self.TaskClass(config=self.config, log=self.log)
-        return task.run(*args)
+
+        try:
+            task.run(*args)
+        except Exception as e:
+            exitStatus = 1
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchedVisitMetricsTask.py:42]: I don't know if we've settled on a convention for "bad" exit statuses?

File: python/lsst/validate/drp/matchedVisitMetricsTask.py [hunk 2]
```diff
@@ -71,9 +85,24 @@ class MatchedVisitMetricsConfig(Config):
...
-    useJointCal = Field(
+    doApplyExternalPhotoCalib = Field(
         dtype=bool, default=False,
-        doc="Whether to use jointcal (or meas_mosaic) to calibrate measurements"
+        doc="Whether to apply external photometric calibration (PhotoCalib)"
+    )
+    externalPhotoCalibName = Field(
+        dtype=str,
+        doc=("Type of external PhotoCalib.  Currently supported are jointcal, "
+             "fgcm, and fgcm_tract."),
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchedVisitMetricsTask.py]: This looks like it should be a `ChoiceField`, to get some automatic validation.

File: python/lsst/validate/drp/matchedVisitMetricsTask.py [hunk 3]
```diff
@@ -71,9 +85,24 @@ class MatchedVisitMetricsConfig(Config):
         dtype=float, default=1.0,
         doc="Match radius (arcseconds)."
     )
-    useJointCal = Field(
+    doApplyExternalPhotoCalib = Field(
         dtype=bool, default=False,
-        doc="Whether to use jointcal (or meas_mosaic) to calibrate measurements"
+        doc="Whether to apply external photometric calibration (PhotoCalib)"
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchedVisitMetricsTask.py]: Maybe reword to: "Whether to apply external photometric calibration via an `lsst.afw.image.PhotoCalib` object. Uses the `externalPhotoCalibName` field to determine what calibration to load."

File: python/lsst/validate/drp/matchedVisitMetricsTask.py [hunk 4]
```diff
@@ -71,9 +85,24 @@ class MatchedVisitMetricsConfig(Config):
...
+    )
+    externalPhotoCalibName = Field(
+        dtype=str,
+        doc=("Type of external PhotoCalib.  Currently supported are jointcal, "
+             "fgcm, and fgcm_tract."),
+        default="jointcal"
+    )
+    doApplyExternalWcs = Field(
+        dtype=bool, default=False,
+        doc="Whether to apply external astrometric calibration (wcs)"
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchedVisitMetricsTask.py]: Maybe reword to: "Whether to apply external astrometric calibration via an `lsst.afw.geom.SkyWcs` object. Uses the `externalSkyWcsName` field to determine what calibration to load."

File: python/lsst/validate/drp/matchedVisitMetricsTask.py [hunk 5]
```diff
@@ -71,9 +85,24 @@ class MatchedVisitMetricsConfig(Config):
...
+    )
+    externalWcsName = Field(
+        dtype=str,
+        doc="Type of external wcs.  Currently supported is jointcal.",
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchedVisitMetricsTask.py]: Another `ChoiceField`.

File: python/lsst/validate/drp/matchedVisitMetricsTask.py [hunk 6]
```diff
@@ -71,9 +85,24 @@ class MatchedVisitMetricsConfig(Config):
...
+        dtype=str,
+        doc=("Type of external PhotoCalib.  Currently supported are jointcal, "
+             "fgcm, and fgcm_tract."),
+        default="jointcal"
+    )
+    doApplyExternalWcs = Field(
+        dtype=bool, default=False,
+        doc="Whether to apply external astrometric calibration (wcs)"
+    )
+    externalWcsName = Field(
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchedVisitMetricsTask.py]: Can you call it `externalSkyWcsName`, since it loads a `SkyWcs` object?

File: python/lsst/validate/drp/matchedVisitMetricsTask.py (hunk 2 above)
Comment (parejkoj) [inline python/lsst/validate/drp/matchedVisitMetricsTask.py]: "Type of external PhotoCalib loaded if `doApplyExternalPhotoCalib=True`. ..."

File: python/lsst/validate/drp/matchedVisitMetricsTask.py [hunk 7]
```diff
@@ -71,9 +85,24 @@ class MatchedVisitMetricsConfig(Config):
...
-        doc="Whether to use jointcal (or meas_mosaic) to calibrate measurements"
+        doc="Whether to apply external photometric calibration (PhotoCalib)"
+    )
+    externalPhotoCalibName = Field(
+        dtype=str,
+        doc=("Type of external PhotoCalib.  Currently supported are jointcal, "
+             "fgcm, and fgcm_tract."),
+        default="jointcal"
+    )
+    doApplyExternalWcs = Field(
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchedVisitMetricsTask.py]: Could you change it to `doApplyExternalSkyWcs`?

File: python/lsst/validate/drp/matchedVisitMetricsTask.py (hunk 5 above)
Comment (parejkoj) [inline python/lsst/validate/drp/matchedVisitMetricsTask.py]: "Type of external SkyWcs loaded if `doApplyExternalSkyWcs=True`. ..."

File: python/lsst/validate/drp/matchedVisitMetricsTask.py [hunk 8]
```diff
@@ -98,12 +127,14 @@ class MatchedVisitMetricsTask(CmdLineTask):
...
-    ignored (but must still be present) and the photometric calibration is
-    retrieved from the `calexp` and the sky positions of sources are loaded
-    directly from the `src` dataset (which is used to obtain raw measurmenets
-    in both cases).
+    `tract`).  When `config.doApplyExternalPhotoCalib` is `True`, the
+    photometric calibration (`photoCalib`) is taken from
+    `config.externalPhotoCalibName`.  Otherwise, the photometric calibration
+    is retrieved from the `calexp`. When `config.doApplyExternalWcs` is
+    `True`, the astrometric calibration (`wcs`) is taken from
+    `config.externalWcsName`.  Otherwise, the astrometric calbration is
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchedVisitMetricsTask.py]: Please note in this paragraph how the Name is converted into the dataset to load (e.g. "the astrometry for dataset `blah` is loaded from `blah_wcs`").

File: python/lsst/validate/drp/matchreduce.py [hunk 9]
```diff
@@ -59,8 +61,16 @@ def build_matched_dataset(repo, dataIds, matchRadius=None, safeSnr=50.,
         Radius for matching. Default is 1 arcsecond.
     safeSnr : `float`, optional
         Minimum median SNR for a match to be considered "safe".
-    useJointCal : `bool`, optional
-        Use jointcal/meas_mosaic outputs to calibrate positions and fluxes.
+    doApplyExternalPhotoCalib : bool, optional
+        Apply external photoCalib to calibrate fluxes.
+        Default is False.
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchreduce.py]: No need to note default values (here and below): they are visible in the docstring from the default kwarg.

File: python/lsst/validate/drp/matchreduce.py [hunk 10]
```diff
@@ -59,8 +61,16 @@ def build_matched_dataset(repo, dataIds, matchRadius=None, safeSnr=50.,
...
+    externalPhotoCalibName : str, optional
+        Type of external `PhotoCalib` to apply.  Currently supported are jointcal,
+        fgcm, and fgcm_tract.  Must be set if doApplyExternalPhotoCalib is True.
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchreduce.py]: Put double-quotes around the supported names.

double-backticks around other kwargs:
```
Must be set if ``doApplyExternalPhotoCalib`` is True.
```

File: python/lsst/validate/drp/matchreduce.py [hunk 11]
```diff
@@ -59,8 +61,16 @@ def build_matched_dataset(repo, dataIds, matchRadius=None, safeSnr=50.,
...
+        Default is None.
+    doApplyExternalWcs : bool, optional
+        Apply external wcs to calibrate positions.
+        Default is False.
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchreduce.py]: You missed `externalWcsName` docstring.

File: python/lsst/validate/drp/matchreduce.py [hunk 12]
```diff
@@ -106,6 +116,11 @@ def build_matched_dataset(repo, dataIds, matchRadius=None, safeSnr=50.,
 
         *Not serialized.*
     """
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchreduce.py:128]: Add a `Raises` block since this could raise `RuntimeError` now.

File: python/lsst/validate/drp/matchreduce.py [hunk 13]
```diff
@@ -116,13 +131,24 @@ def build_matched_dataset(repo, dataIds, matchRadius=None, safeSnr=50.,
                                description='Filter name')
 
     # Record important configuration
-    blob['useJointCal'] = Datum(quantity=useJointCal,
-                                description='Whether jointcal/meas_mosaic calibrations were used')
+    blob['doApplyExternalPhotoCalib'] = Datum(quantity=doApplyExternalPhotoCalib,
+                                              description=('Whether external photometric '
+                                                           'calibrations were used'))
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchreduce.py]: These descriptions are sentences, so should have a period at the end.

File: python/lsst/validate/drp/matchreduce.py [hunk 14]
```diff
@@ -116,13 +131,24 @@ def build_matched_dataset(repo, dataIds, matchRadius=None, safeSnr=50.,
...
+    blob['externalPhotoCalibName'] = Datum(quantity=externalPhotoCalibName,
+                                           description='Name of external PhotoCalib used')
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchreduce.py]: "Name of external PhotoCalib dataset used"? (same for wcs below)

File: python/lsst/validate/drp/matchreduce.py [hunk 15]
```diff
@@ -160,6 +198,12 @@ def _loadAndMatchCatalogs(repo, dataIds, matchRadius,
     matches : `lsst.afw.table.GroupView`
         A GroupView of the matched sources.
     """
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchreduce.py:216]: Need a `Raises` here too.

File: python/lsst/validate/drp/matchreduce.py [hunk 16]
```diff
@@ -146,9 +174,19 @@ def _loadAndMatchCatalogs(repo, dataIds, matchRadius,
         calibration.
     matchRadius :  `lsst.geom.Angle`, optional
         Radius for matching. Default is 1 arcsecond.
-    useJointCal : `bool`, optional
-        Use jointcal outputs to calibrate positions and fluxes instead of
-        meas_mosaic.
+    doApplyExternalPhotoCalib : bool, optional
+        Apply external photoCalib to calibrate fluxes.
+        Default is False.
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchreduce.py]: No need for defaults.

File: python/lsst/validate/drp/matchreduce.py [hunk 17]
```diff
@@ -212,29 +256,21 @@ def _loadAndMatchCatalogs(repo, dataIds, matchRadius,
...
             except (FitsError, dafPersist.NoResults) as e:
                 print(e)
-                print("Could not open updated WCS for ", vId)
+                print("Could not open external photometric calibration for ", vId)
                 print("Skipping this dataId.")
                 continue
         else:
             try:
-                photoCalib = butler.get("calexp_photoCalib", vId)
+                photoCalib = butler.get('calexp_photoCalib', vId)
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchreduce.py]: Why change to single quotes here?

File: python/lsst/validate/drp/matchreduce.py [hunk 18]
```diff
@@ -212,29 +256,21 @@ def _loadAndMatchCatalogs(repo, dataIds, matchRadius,
...
 
-        if useJointCal:
-            try:
-                photoCalib = butler.get("jointcal_photoCalib", vId)
-            except (FitsError, dafPersist.NoResults) as e:
-                print(e)
-                print("Could not open photometric calibration for ", vId)
-                print("Skipping this dataId.")
-                continue
+        if doApplyExternalPhotoCalib:
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchreduce.py]: At this point, I think we should pull these `if` statements into separate methods `loadPhotoCalib`, `loadSkyWcs`. That will help clean this overly-long method up.

File: python/lsst/validate/drp/matchreduce.py [hunk 19]
```diff
@@ -212,29 +256,21 @@ def _loadAndMatchCatalogs(repo, dataIds, matchRadius,
...
             try:
-                wcs = butler.get("jointcal_wcs", vId)
+                photoCalib = butler.get('%s_photoCalib' % (externalPhotoCalibName), vId)
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchreduce.py]: Use an f-string here: `f"{externalPhotoCalibName}_photoCalib"`

File: python/lsst/validate/drp/matchreduce.py [hunk 20]
```diff
@@ -249,9 +285,18 @@ def _loadAndMatchCatalogs(repo, dataIds, matchRadius,
                 print("Skipping this dataId.")
                 continue
 
-        # We don't want to put this above the first "if useJointCal block"
-        # because we need to use the first `butler.get` above to quickly
-        # catch data IDs with no usable outputs.
+        if doApplyExternalWcs:
+            try:
+                wcs = butler.get("%s_wcs" % (externalWcsName), vId)
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchreduce.py]: f-string!

File: python/lsst/validate/drp/matchreduce.py [hunk 21]
```diff
@@ -249,9 +285,18 @@ def _loadAndMatchCatalogs(repo, dataIds, matchRadius,
...
+            except (FitsError, dafPersist.NoResults) as e:
+                print(e)
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchreduce.py]: Do we have a logger here? I know the `print`s were pre-existing, but it would be good to replace them with `log.warn`s.

File: python/lsst/validate/drp/matchreduce.py [hunk 22]
```diff
@@ -268,7 +313,7 @@ def _loadAndMatchCatalogs(repo, dataIds, matchRadius,
         tmpCat['base_PsfFlux_snr'][:] = tmpCat['base_PsfFlux_instFlux'] \
             / tmpCat['base_PsfFlux_instFluxErr']
 
-        if useJointCal:
+        if doApplyExternalWcs:
             for record in tmpCat:
                 record.updateCoord(wcs)
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchreduce.py]: I know you didn't touch this line, but please replace this loop with `afw.table.wcsUtils.updateSourceCoords(wcs, tmpCat)`, which does the loop in C++.

File: python/lsst/validate/drp/validate.py [hunk 23]
```diff
@@ -282,22 +285,42 @@ def runOneFilter(repo, visitDataIds, metrics, brightSnr=100,
...
+        Default is False.
+    externalWcsName : str, optional
+        Type of external `wcs` to apply.  Currently supported is jointcal.
+        Must be set if doApplyExternalWcs is True.  Default is None.
     skipTEx : bool, optional
         Skip TEx calculations (useful for older catalogs that don't have
         PsfShape measurements).
     verbose : bool, optional
         Output additional information on the analysis steps.
     """
```
Comment (parejkoj) [inline python/lsst/validate/drp/validate.py:313]: Please add `Raises` and cleanup the docs as in the other methods.

**Review by parejkoj — CHANGES_REQUESTED** [review]
Thanks for adding the exitStatus to `__call__`. Please squash the two commits related to that, since they go together.

File: python/lsst/validate/drp/matchedVisitMetricsTask.py (hunk 1 above)
Comment (erykoff) [reply python/lsst/validate/drp/matchedVisitMetricsTask.py:42]: When you use the old-form code (no proper exit status) it says "Assuming exit status of 0" or something like that, which I think means it's assuming 0 is good.

File: python/lsst/validate/drp/matchedVisitMetricsTask.py (hunk 2 above)
Comment (erykoff) [reply python/lsst/validate/drp/matchedVisitMetricsTask.py]: Thanks, I didn't know about the `ChoiceField`.

File: python/lsst/validate/drp/matchreduce.py (hunk 17 above)
Comment (erykoff) [reply python/lsst/validate/drp/matchreduce.py]: Consistency with the other `butler.get` ...

File: python/lsst/validate/drp/matchreduce.py (hunk 21 above)
Comment (erykoff) [reply python/lsst/validate/drp/matchreduce.py]: I don't think we have a logger, and I'm not sure how to get one ... I hope it's okay if I punt on this as out of scope?

Comment (parejkoj) [reply python/lsst/validate/drp/matchreduce.py]: Right this is all outside of a Task. Getting this to use a proper logger should happen, but it is out of scope for you.

File: python/lsst/validate/drp/matchedVisitMetricsTask.py [hunk 24]
```diff
@@ -71,9 +84,36 @@ class MatchedVisitMetricsConfig(Config):
         dtype=float, default=1.0,
         doc="Match radius (arcseconds)."
     )
-    useJointCal = Field(
+    doApplyExternalPhotoCalib = Field(
+        dtype=bool, default=False,
+        doc=("Whether to apply external photometric calibration via an "
+             "`lsst.afw.image.PhotoCalib` object.  Uses the "
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchedVisitMetricsTask.py:90]: two spaces? ;-)

https://xkcd.com/1285/

Also, I don't think these docstrings need to be wrapped at 80 characters. At least, I hope they don't.

File: python/lsst/validate/drp/matchreduce.py [hunk 25]
```diff
@@ -409,3 +437,86 @@ def fullFilter(cat):
...
+
+
+def _loadPhotoCalib(butler, dataId, doApplyExternalPhotoCalib, externalPhotoCalibName):
+    """
+    Load a photoCalib object.
+
+    Parameters
+    ----------
+    butler: `lsst.daf.persistence.Butler`
+    dataId: Butler dataId
```
Comment (parejkoj) [inline python/lsst/validate/drp/matchreduce.py]: Docstrings for these two (and below)? Is the "Butler dataId" a `dict` or an `lsst.daf.persistence.ButlerDataRef`?

File: python/lsst/validate/drp/validate.py [hunk 26]
```diff
@@ -283,24 +286,50 @@ def runOneFilter(repo, visitDataIds, metrics, brightSnr=100,
         Specify the beginning filename for output files.
     filterName : str, optional
         Name of the filter (bandpass).
-    useJointCal : bool, optional
-        Use jointcal/meas_mosaic outputs to calibrate positions and fluxes.
+    doApplyExternalPhotoCalib : bool, optional
+        Apply external photoCalib to calibrate fluxes.
+        Default is False.
```
Comment (parejkoj) [inline python/lsst/validate/drp/validate.py]: Don't need to specify defaults here (and below).

**Review by parejkoj — APPROVED** [review]
A handful of other comments here on some docs. Clean those up and you're good to merge.
